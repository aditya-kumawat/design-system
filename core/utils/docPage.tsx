// @ts-nocheck

import * as React from 'react';
import { Title, Subtitle, Heading, Subheading, Props, Description, Preview, Source, PropsProps } from '@storybook/addon-docs/blocks';
import { renderToStaticMarkup } from 'react-dom/server';
import reactElementToJSXString from 'react-element-to-jsx-string';
import { html as beautifyHTML } from 'js-beautify';
import SyntaxHighlighter from 'react-syntax-highlighter';
import { docco, dark, a11yDark, githubGist, dracula, vs2015 } from 'react-syntax-highlighter/dist/esm/styles/hljs';
import * as DS from '@/';
import { Button, Card, TabsWrapper, Tab } from '@/';
import vsDark from 'prism-react-renderer/themes/vsDark';

import {
  LiveProvider,
  LiveEditor,
  LiveError,
  LivePreview,
  withLive
} from 'react-live';

export interface Example {
  title: string;
  description?: string;
  imports: string[];
  component: React.ReactNode;
}

interface DocPageProps {
  title?: string;
  description?: React.ReactNode;
  props?: PropsProps;
  customCode?: string;
}

const beautifyHTMLOptions = {
  indent_size: 2,
  wrap_line_length: 0,
  preserve_newlines: true,
  jslint_happy: true,
  end_with_newline: false,
  indent_inner_html: true
};

const beautifyJSXOptions = {
  ...beautifyHTMLOptions,
  e4x: true
};

const JSXtoStringOptions = {
  filterProps: (val: any, key: string) => {
    if (!val) return false;
    if (val && val.name === 'actionHandler') return false;
    // if(val && typeof val === 'function') return false;
    return true;
  },
  showFunctions: true,
  functionValue: _fn => _ => { },
  // maxInlineAttributesLineLength: 10
};

const copyCode = (val: string) => navigator.clipboard.writeText(val);

const importsToStr = (imports: string[]): string => `import { ${imports.join(', ')} } from 'design-system';`;

const CopyComp = props => {
  const { onClick } = props;
  return (
    <div
      style={{
        position: 'absolute',
        right: '0',
        top: '0',
        zIndex: 10
      }}
    >
      <Button size="small" style={{ borderRadius: '0' }} appearance="basic" onClick={onClick}>Copy</Button>
    </div>
  );
};

const renderCodeBlock = (val: string) => (
  <pre style={{ position: 'relative', fontFamily: 'monospace', fontSize: '13px', background: '#f8f8f8' }}>
    <CopyComp onClick={() => copyCode(val)} />
    <SyntaxHighlighter language="javascript" style={vs2015} showLineNumbers={true}>
      {val}
    </SyntaxHighlighter>
  </pre>
);

const StoryComp = props => {
  const {
    customCode
  } = props;

  const { storyId } = __STORYBOOK_STORY_STORE__.getSelection();
  const story = __STORYBOOK_STORY_STORE__.fromId(storyId);
  const comp = story.storyFn().props.children.props.children.props.children;
  const sp = story.parameters;
  const imports = [sp.component, ...(sp.subcomponents ? Object.values(sp.subcomponents) : [])];

  const importString = `// ${importsToStr(imports.map(i => i.displayName))}`;
  const jsx = customCode ? customCode : `${beautifyHTML(reactElementToJSXString(comp, JSXtoStringOptions), beautifyJSXOptions)}`;
  const html = beautifyHTML(renderToStaticMarkup(comp), beautifyHTMLOptions);

  const [activeTab, setActiveTab] = React.useState<number>(0);
  const [jsxCode, setJsxCode] = React.useState<string>(`${importString}\n\n${jsx}`);
  const [htmlCode, setHtmlCode] = React.useState<string>(`${html}`);

  const importScope = imports.reduce((out, curr) => {
    out[curr.displayName] = curr;
    return out;
  }, {});

  const TabsWrap = withLive(({ live }) => {
    const {
      error,
      element: Element,
    } = live;

    React.useEffect(() => {
      if (!error) {
        const editor = document.querySelector('.npm__react-simple-code-editor__textarea');

        if (editor) {
          const jsxValue = editor.value;
          setJsxCode(jsxValue);

          try {
            const htmlValue = beautifyHTML(renderToStaticMarkup(<Element />), beautifyHTMLOptions);
            setHtmlCode(htmlValue);
          } catch (e) { }
        }
      }
    }, [live]);

    return (
      <></>
    );
  });

  return (
    <>
      <Card
        shadow="light"
      >
        <LiveProvider code={jsxCode} scope={{ ...DS, ...importScope }}>
          <Preview
            style={{
              marginTop: '0px',
              marginBottom: '0px'
            }}
            withSource="none"
            withToolbar={true}
          >
            <LivePreview />
          </Preview>
          <TabsWrap />
          <TabsWrapper
            activeTab={activeTab}
            onTabChange={tab => setActiveTab(tab)}
          >
            <Tab label={'React'}>
              <div style={{ position: 'relative' }}>
                <CopyComp
                  onClick={() => {
                    const editor = document.querySelector('.npm__react-simple-code-editor__textarea');
                    if (editor) copyCode(editor.value);
                  }}
                />
                <LiveEditor theme={vsDark} />
                <LiveError />
              </div>
            </Tab>
            <Tab label={'HTML'}>
              {renderCodeBlock(htmlCode)}
            </Tab>
          </TabsWrapper>
        </LiveProvider>
      </Card>
    </>
  );
};

export const docPage = () => {
  const { storyId } = __STORYBOOK_STORY_STORE__.getSelection();
  const story = __STORYBOOK_STORY_STORE__.fromId(storyId);
  const comp = story.storyFn().props.children.props.children.props.children;
  const sp = story.parameters;

  const {
    title,
    description,
    props: propsAttr,
    customCode,
    noStory
  } = sp.docs.docPage;

  const separatorIndex = title?.indexOf('|');

  return (
    <>
      <Title>{title && separatorIndex !== -1 ? title.slice(separatorIndex + 1) : title}</Title>
      <br />

      <Description>
        {description}
      </Description>
      <br />

      {!noStory && (
        <>
          <Heading>Story</Heading>
          <StoryComp key={storyId} customCode={customCode} />
          <br />
        </>
      )}

      <Heading>PropTable</Heading>
      <Props {...propsAttr} />
    </>
  );
};

export default docPage;
