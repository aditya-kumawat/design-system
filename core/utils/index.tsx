import * as positionHelper from './positionHelper';
import PopperWrapper from './PopperWrapper';

export { positionHelper,  PopperWrapper };
export * from './css';
