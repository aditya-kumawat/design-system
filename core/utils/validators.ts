export const dateValidator = (format: string, val: string): boolean => {
  const validate = (date: number, month: number, year: number): boolean => {
    // Check the ranges of month and year
    if (year < 1000 || year > 3000 || month == 0 || month > 12)
      return false;

    var monthLength = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];

    // Adjust for leap years
    if (year % 400 == 0 || (year % 100 != 0 && year % 4 == 0))
      monthLength[1] = 29;

    // Check the range of the day
    return date > 0 && date <= monthLength[month - 1];
  }

  switch (format) {
    case 'dd/mm/yy':
      var p = val.split('/');
      var date = +p[0];
      var month = +p[1];
      var year = +p[2];
      return validate(date, month, year);

    case 'mm/dd/yy':
      var p = val.split('/');
      var date = +p[1];
      var month = +p[0];
      var year = +p[2];
      return validate(date, month, year);

    case 'yy/mm/dd':
      var p = val.split('/');
      var date = +p[2];
      var month = +p[1];
      var year = +p[0];
      return validate(date, month, year);

    case 'dd-mm-yy':
      var p = val.split('-');
      var date = +p[0];
      var month = +p[1];
      var year = +p[2];
      return validate(date, month, year);

    case 'mm-dd-yy':
      var p = val.split('-');
      var date = +p[1];
      var month = +p[0];
      var year = +p[2];
      return validate(date, month, year);

    case 'yy-mm-dd':
      var p = val.split('-');
      var date = +p[2];
      var month = +p[1];
      var year = +p[0];
      return validate(date, month, year);

    default:
      return false;
  }
}

const e: Record<string, any> = {
  date: dateValidator
}
export default e;