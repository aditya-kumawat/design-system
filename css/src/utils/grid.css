.Row {
  display: flex;
  flex-wrap: wrap;
}

.Col {
  position: relative;
  width: 100%;
}

.RowGroup--1 > .Col {
  flex: 0 0;
	flex-basis: calc((100% / 1));
  max-width: calc((100% / 1));
}

.RowGroup--2 > .Col {
  flex: 0 0;
	flex-basis: calc((100% / 2));
  max-width: calc((100% / 2));
}

.RowGroup--3 > .Col {
  flex: 0 0;
	flex-basis: calc((100% / 3));
  max-width: calc((100% / 3));
}

.RowGroup--4 > .Col {
  flex: 0 0;
	flex-basis: calc((100% / 4));
  max-width: calc((100% / 4));
}

.RowGroup--5 > .Col {
  flex: 0 0;
	flex-basis: calc((100% / 5));
  max-width: calc((100% / 5));
}

.RowGroup--6 > .Col {
  flex: 0 0;
	flex-basis: calc((100% / 6));
  max-width: calc((100% / 6));
}

.Col--auto {
  flex: 0 0 auto;
  width: auto;
  max-width: 100%;
}

.Col--1 {
  flex: 0 0;
	flex-basis: calc((100% / (var(--grid-columns) / 1)));
  max-width: calc((100% / (var(--grid-columns) / 1)));
}

.Col--2 {
  flex: 0 0;
  flex-basis: calc((100% / (var(--grid-columns) / 2)));
  max-width: calc((100% / (var(--grid-columns) / 2)));
}

.Col--3 {
  flex: 0 0 25%;
  max-width: calc((100% / (var(--grid-columns) / 3)));
}

.Col--4 {
  flex: 0 0;
	flex-basis: calc((100% / (var(--grid-columns) / 4)));
  max-width: calc((100% / (var(--grid-columns) / 4)));
}

.Col--5 {
  flex: 0 0;
	flex-basis: calc((100% / (var(--grid-columns) / 5)));
  max-width: calc((100% / (var(--grid-columns) / 5)));
}

.Col--6 {
  flex: 0 0;
	flex-basis: calc((100% / (var(--grid-columns) / 6)));
  max-width: calc((100% / (var(--grid-columns) / 6)));
}

.Col--7 {
  flex: 0 0;
	flex-basis: calc((100% / (var(--grid-columns) / 7)));
  max-width: calc((100% / (var(--grid-columns) / 7)));
}

.Col--8 {
  flex: 0 0;
	flex-basis: calc((100% / (var(--grid-columns) / 8)));
  max-width: calc((100% / (var(--grid-columns) / 8)));
}

.Col--9 {
  flex: 0 0;
	flex-basis: calc((100% / (var(--grid-columns) / 9)));
  max-width: calc((100% / (var(--grid-columns) / 9)));
}

.Col--10 {
  flex: 0 0;
	flex-basis: calc((100% / (var(--grid-columns) / 10)));
  max-width: calc((100% / (var(--grid-columns) / 10)));
}

.Col--11 {
  flex: 0 0;
	flex-basis: calc((100% / (var(--grid-columns) / 11)));
  max-width: calc((100% / (var(--grid-columns) / 11)));
}

.Col--12 {
  flex: 0 0;
	flex-basis: calc((100% / (var(--grid-columns) / 12)));
  max-width: calc((100% / (var(--grid-columns) / 12)));
}

/* if the viewport is between 0 to 575 pixels wide. */
@media (max-width: 575px) {
  .RowGroup--xs-1 > .Col {
    flex: 0 0;
		flex-basis: calc((100% / 1));
    max-width: calc((100% / 1));
  }

  .RowGroup--xs-2 > .Col {
    flex: 0 0;
		flex-basis: calc((100% / 2));
    max-width: calc((100% / 2));
  }

  .RowGroup--xs-3 > .Col {
    flex: 0 0;
		flex-basis: calc((100% / 3));
    max-width: calc((100% / 3));
  }

  .RowGroup--xs-4 > .Col {
    flex: 0 0;
		flex-basis: calc((100% / 4));
    max-width: calc((100% / 4));
  }

  .RowGroup--xs-5 > .Col {
    flex: 0 0;
		flex-basis: calc((100% / 5));
    max-width: calc((100% / 5));
  }

  .RowGroup--xs-6 > .Col {
    flex: 0 0;
		flex-basis: calc((100% / 6));
    max-width: calc((100% / 6));
  }

  .Col--xs {
    flex-basis: 0;
    flex-grow: 1;
    max-width: 100%;
  }

  .Col--xs-auto {
    flex: 0 0 auto;
    width: auto;
    max-width: 100%;
  }

  .Col--xs-1 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 1)));
    max-width: calc((100% / (var(--grid-columns) / 1)));
  }

  .Col--xs-2 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 2)));
    max-width: calc((100% / (var(--grid-columns) / 2)));
  }

  .Col--xs-3 {
    flex: 0 0 25%;
    max-width: calc((100% / (var(--grid-columns) / 3)));
  }

  .Col--xs-4 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 4)));
    max-width: calc((100% / (var(--grid-columns) / 4)));
  }

  .Col--xs-5 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 5)));
    max-width: calc((100% / (var(--grid-columns) / 5)));
  }

  .Col--xs-6 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 6)));
    max-width: calc((100% / (var(--grid-columns) / 6)));
  }

  .Col--xs-7 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 7)));
    max-width: calc((100% / (var(--grid-columns) / 7)));
  }

  .Col--xs-8 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 8)));
    max-width: calc((100% / (var(--grid-columns) / 8)));
  }

  .Col--xs-9 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 9)));
    max-width: calc((100% / (var(--grid-columns) / 9)));
  }

  .Col--xs-10 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 10)));
    max-width: calc((100% / (var(--grid-columns) / 10)));
  }

  .Col--xs-11 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 11)));
    max-width: calc((100% / (var(--grid-columns) / 11)));
  }

  .Col--xs-12 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 12)));
    max-width: calc((100% / (var(--grid-columns) / 12)));
  }
}

/* if the viewport is between 576 and 767 pixels wide. */
@media (min-width: 576px) and (max-width: 767px) {
  .RowGroup--s-1 > .Col {
    flex: 0 0;
		flex-basis: calc((100% / 1));
    max-width: calc((100% / 1));
  }

  .RowGroup--s-2 > .Col {
    flex: 0 0;
		flex-basis: calc((100% / 2));
    max-width: calc((100% / 2));
  }

  .RowGroup--s-3 > .Col {
    flex: 0 0;
		flex-basis: calc((100% / 3));
    max-width: calc((100% / 3));
  }

  .RowGroup--s-4 > .Col {
    flex: 0 0;
		flex-basis: calc((100% / 4));
    max-width: calc((100% / 4));
  }

  .RowGroup--s-5 > .Col {
    flex: 0 0;
		flex-basis: calc((100% / 5));
    max-width: calc((100% / 5));
  }

  .RowGroup--s-6 > .Col {
    flex: 0 0;
		flex-basis: calc((100% / 6));
    max-width: calc((100% / 6));
  }

  .Col--s-auto {
    flex: 0 0 auto;
    width: auto;
    max-width: 100%;
  }

  .Col--s-1 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 1)));
    max-width: calc((100% / (var(--grid-columns) / 1)));
  }

  .Col--s-2 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 2)));
    max-width: calc((100% / (var(--grid-columns) / 2)));
  }

  .Col--s-3 {
    flex: 0 0 25%;
    max-width: calc((100% / (var(--grid-columns) / 3)));
  }

  .Col--s-4 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 4)));
    max-width: calc((100% / (var(--grid-columns) / 4)));
  }

  .Col--s-5 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 5)));
    max-width: calc((100% / (var(--grid-columns) / 5)));
  }

  .Col--s-6 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 6)));
    max-width: calc((100% / (var(--grid-columns) / 6)));
  }

  .Col--s-7 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 7)));
    max-width: calc((100% / (var(--grid-columns) / 7)));
  }

  .Col--s-8 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 8)));
    max-width: calc((100% / (var(--grid-columns) / 8)));
  }

  .Col--s-9 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 9)));
    max-width: calc((100% / (var(--grid-columns) / 9)));
  }

  .Col--s-10 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 10)));
    max-width: calc((100% / (var(--grid-columns) / 10)));
  }

  .Col--s-11 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 11)));
    max-width: calc((100% / (var(--grid-columns) / 11)));
  }

  .Col--s-12 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 12)));
    max-width: calc((100% / (var(--grid-columns) / 12)));
  }
}

/* if the viewport is between 768 and 991 pixels wide. */
@media (min-width: 768px) and (max-width: 991px) {
  .RowGroup--m-1 > .Col {
    flex: 0 0;
		flex-basis: calc((100% / 1));
    max-width: calc((100% / 1));
  }

  .RowGroup--m-2 > .Col {
    flex: 0 0;
		flex-basis: calc((100% / 2));
    max-width: calc((100% / 2));
  }

  .RowGroup--m-3 > .Col {
    flex: 0 0;
		flex-basis: calc((100% / 3));
    max-width: calc((100% / 3));
  }

  .RowGroup--m-4 > .Col {
    flex: 0 0;
		flex-basis: calc((100% / 4));
    max-width: calc((100% / 4));
  }

  .RowGroup--m-5 > .Col {
    flex: 0 0;
		flex-basis: calc((100% / 5));
    max-width: calc((100% / 5));
  }

  .RowGroup--m-6 > .Col {
    flex: 0 0;
		flex-basis: calc((100% / 6));
    max-width: calc((100% / 6));
  }

  .Col--m-auto {
    flex: 0 0 auto;
    width: auto;
    max-width: 100%;
  }

  .Col--m-1 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 1)));
    max-width: calc((100% / (var(--grid-columns) / 1)));
  }

  .Col--m-2 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 2)));
    max-width: calc((100% / (var(--grid-columns) / 2)));
  }

  .Col--m-3 {
    flex: 0 0 25%;
    max-width: calc((100% / (var(--grid-columns) / 3)));
  }

  .Col--m-4 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 4)));
    max-width: calc((100% / (var(--grid-columns) / 4)));
  }

  .Col--m-5 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 5)));
    max-width: calc((100% / (var(--grid-columns) / 5)));
  }

  .Col--m-6 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 6)));
    max-width: calc((100% / (var(--grid-columns) / 6)));
  }

  .Col--m-7 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 7)));
    max-width: calc((100% / (var(--grid-columns) / 7)));
  }

  .Col--m-8 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 8)));
    max-width: calc((100% / (var(--grid-columns) / 8)));
  }

  .Col--m-9 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 9)));
    max-width: calc((100% / (var(--grid-columns) / 9)));
  }

  .Col--m-10 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 10)));
    max-width: calc((100% / (var(--grid-columns) / 10)));
  }

  .Col--m-11 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 11)));
    max-width: calc((100% / (var(--grid-columns) / 11)));
  }

  .Col--m-12 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 12)));
    max-width: calc((100% / (var(--grid-columns) / 12)));
  }
}

/* if the viewport is between 992 and 1199 pixels wide. */
@media (min-width: 992px) and (max-width: 1199px) {
  .RowGroup--l-1 > .Col {
    flex: 0 0;
		flex-basis: calc((100% / 1));
    max-width: calc((100% / 1));
  }

  .RowGroup--l-2 > .Col {
    flex: 0 0;
		flex-basis: calc((100% / 2));
    max-width: calc((100% / 2));
  }

  .RowGroup--l-3 > .Col {
    flex: 0 0;
		flex-basis: calc((100% / 3));
    max-width: calc((100% / 3));
  }

  .RowGroup--l-4 > .Col {
    flex: 0 0;
		flex-basis: calc((100% / 4));
    max-width: calc((100% / 4));
  }

  .RowGroup--l-5 > .Col {
    flex: 0 0;
		flex-basis: calc((100% / 5));
    max-width: calc((100% / 5));
  }

  .RowGroup--l-6 > .Col {
    flex: 0 0;
		flex-basis: calc((100% / 6));
    max-width: calc((100% / 6));
  }

  .Col--l-auto {
    flex: 0 0 auto;
    width: auto;
    max-width: 100%;
  }

  .Col--l-1 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 1)));
    max-width: calc((100% / (var(--grid-columns) / 1)));
  }

  .Col--l-2 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 2)));
    max-width: calc((100% / (var(--grid-columns) / 2)));
  }

  .Col--l-3 {
    flex: 0 0 25%;
    max-width: calc((100% / (var(--grid-columns) / 3)));
  }

  .Col--l-4 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 4)));
    max-width: calc((100% / (var(--grid-columns) / 4)));
  }

  .Col--l-5 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 5)));
    max-width: calc((100% / (var(--grid-columns) / 5)));
  }

  .Col--l-6 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 6)));
    max-width: calc((100% / (var(--grid-columns) / 6)));
  }

  .Col--l-7 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 7)));
    max-width: calc((100% / (var(--grid-columns) / 7)));
  }

  .Col--l-8 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 8)));
    max-width: calc((100% / (var(--grid-columns) / 8)));
  }

  .Col--l-9 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 9)));
    max-width: calc((100% / (var(--grid-columns) / 9)));
  }

  .Col--l-10 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 10)));
    max-width: calc((100% / (var(--grid-columns) / 10)));
  }

  .Col--l-11 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 11)));
    max-width: calc((100% / (var(--grid-columns) / 11)));
  }

  .Col--l-12 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 12)));
    max-width: calc((100% / (var(--grid-columns) / 12)));
  }
}

/* if the viewport is 1200 pixels wide or wider. */
@media (min-width: 1200px) {
  .RowGroup--xl-1 > .Col {
    flex: 0 0;
		flex-basis: calc((100% / 1));
    max-width: calc((100% / 1));
  }

  .RowGroup--xl-2 > .Col {
    flex: 0 0;
		flex-basis: calc((100% / 2));
    max-width: calc((100% / 2));
  }

  .RowGroup--xl-3 > .Col {
    flex: 0 0;
		flex-basis: calc((100% / 3));
    max-width: calc((100% / 3));
  }

  .RowGroup--xl-4 > .Col {
    flex: 0 0;
		flex-basis: calc((100% / 4));
    max-width: calc((100% / 4));
  }

  .RowGroup--xl-5 > .Col {
    flex: 0 0;
		flex-basis: calc((100% / 5));
    max-width: calc((100% / 5));
  }

  .RowGroup--xl-6 > .Col {
    flex: 0 0;
		flex-basis: calc((100% / 6));
    max-width: calc((100% / 6));
  }

  .Col--xl-auto {
    flex: 0 0 auto;
    width: auto;
    max-width: 100%;
  }

  .Col--xl-1 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 1)));
    max-width: calc((100% / (var(--grid-columns) / 1)));
  }

  .Col--xl-2 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 2)));
    max-width: calc((100% / (var(--grid-columns) / 2)));
  }

  .Col--xl-3 {
    flex: 0 0 25%;
    max-width: calc((100% / (var(--grid-columns) / 3)));
  }

  .Col--xl-4 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 4)));
    max-width: calc((100% / (var(--grid-columns) / 4)));
  }

  .Col--xl-5 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 5)));
    max-width: calc((100% / (var(--grid-columns) / 5)));
  }

  .Col--xl-6 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 6)));
    max-width: calc((100% / (var(--grid-columns) / 6)));
  }

  .Col--xl-7 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 7)));
    max-width: calc((100% / (var(--grid-columns) / 7)));
  }

  .Col--xl-8 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 8)));
    max-width: calc((100% / (var(--grid-columns) / 8)));
  }

  .Col--xl-9 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 9)));
    max-width: calc((100% / (var(--grid-columns) / 9)));
  }

  .Col--xl-10 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 10)));
    max-width: calc((100% / (var(--grid-columns) / 10)));
  }

  .Col--xl-11 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 11)));
    max-width: calc((100% / (var(--grid-columns) / 11)));
  }

  .Col--xl-12 {
    flex: 0 0;
		flex-basis: calc((100% / (var(--grid-columns) / 12)));
    max-width: calc((100% / (var(--grid-columns) / 12)));
  }
}
