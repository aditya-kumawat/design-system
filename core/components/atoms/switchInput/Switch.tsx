import * as React from 'react';
import classNames from 'classnames';
import { BaseProps, extractBaseProps } from '@/utils/types';

export type Size = 'regular' | 'tiny' | 'large';
export type Appearance = 'primary' | 'alert' | 'success' | 'warning';

type MouseEvent = React.ChangeEvent<HTMLInputElement>;

export interface SwitchProps extends BaseProps {
  /**
   * Size of `Switch`
   * @default "regular"
   */
  size?: Size;
  /**
   * Color of `Switch`
   *
   * **Soon to be deprecated**
   */
  appearance?: Appearance;
  /**
   * Default value of checked (Used in case of uncontrolled `Switch`)
   */
  defaultChecked?: boolean;
  /**
   * Denotes Selection (Used in case of controlled `Switch`)
   */
  checked?: boolean;
  /**
   * Disables the `Switch`, making it unable to be pressed
   */
  disabled?: boolean;
  /**
   * Name of `Switch`
   */
  name?: string;
  /**
   * Value of `Switch`
   */
  value?: string;
  /**
   * Callback function called when `Switch` is toggled
   */
  onChange?: (event: MouseEvent, selected: boolean) => void;
}

/**
 * ######Switch has two types:
 *  - [Controlled Switch](https://reactjs.org/docs/forms.html#controlled-components)
 *  - [Uncontrolled Switch](https://reactjs.org/docs/uncontrolled-components.html)
 */
export const Switch = React.forwardRef<HTMLInputElement, SwitchProps>((props, ref) => {
  const {
    size = 'regular',
    defaultChecked,
    disabled,
    onChange,
    name,
    value,
    className
  } = props;

  const baseProps = extractBaseProps(props);

  const [checked, setChecked] = React.useState(props.checked === undefined ? defaultChecked : props.checked);

  React.useEffect(() => {
    if (props.checked !== undefined) setChecked(props.checked);
  }, [props.checked]);

  const SwitchClass = classNames({
    ['Switch']: true,
    ['Switch--disabled']: disabled,
    [`Switch--${size}`]: size,
  }, className);

  const SwitchWrapper = classNames({
    ['Switch-wrapper']: true,
    ['Switch-wrapper--disabled']: disabled,
    [`Switch-wrapper--${size}`]: size,
    ['Switch-wrapper--checked']: checked,
    ['Switch-wrapper--checkedDisabled']: checked && disabled,
  });

  const onChangeHandler = (event: MouseEvent) => {
    if (props.checked === undefined) setChecked(!checked);
    if (onChange) onChange(event, !checked);
  };

  return (
    <div className={SwitchClass}>
      <input
        {...baseProps}
        type="checkbox"
        defaultChecked={defaultChecked}
        disabled={disabled}
        onChange={onChangeHandler}
        checked={checked}
        ref={ref}
        name={name}
        value={value}
        className="Switch-input"
      />
      <span className={SwitchWrapper} />
    </div>
  );
});

Switch.displayName = 'Switch';

export default Switch;
