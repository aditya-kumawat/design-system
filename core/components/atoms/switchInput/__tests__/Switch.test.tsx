import * as React from 'react';
import { shallow } from 'enzyme';
import { testHelper, filterUndefined, valueHelper, testMessageHelper } from '@/utils/testHelper';
import Switch, { SwitchProps as Props } from '../Switch';

export const size = ['tiny', 'regular'];
const BooleanValue = [true, false];

describe('Switch component', () => {
  const mapper: Record<string, any> = {
    checked: valueHelper(true, { required: true }),
    disabled: valueHelper(BooleanValue, { required: true, iterate: true }),
  };

  const testFunc = (props: Record<string, any>): void => {
    const attr = filterUndefined(props) as Props;
    it(testMessageHelper(attr), () => {
      const tree = shallow(
        <Switch
          {...attr}
        />
      );
      expect(tree).toMatchSnapshot();
    });
  };
  testHelper(mapper, testFunc);
});

describe('Switch component', () => {
  const mapper: Record<string, any> = {
    size: valueHelper(size, { iterate: true }),
    checked: valueHelper(true, { required: true }),
  };

  const testFunc = (props: Record<string, any>): void => {
    const attr = filterUndefined(props) as Props;
    it(testMessageHelper(attr), () => {
      const tree = shallow(
        <Switch
          {...attr}
        />
      );
      expect(tree).toMatchSnapshot();
    });
  };
  testHelper(mapper, testFunc);
});
