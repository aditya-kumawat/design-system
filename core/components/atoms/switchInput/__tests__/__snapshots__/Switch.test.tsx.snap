// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch component 
  checked: true
 1`] = `
<div
  className="Switch Switch--regular"
>
  <input
    checked={true}
    className="Switch-input"
    onChange={[Function]}
    type="checkbox"
  />
  <span
    className="Switch-wrapper Switch-wrapper--regular Switch-wrapper--checked"
  />
</div>
`;

exports[`Switch component 
  checked: true, disabled: false
 1`] = `
<div
  className="Switch Switch--regular"
>
  <input
    checked={true}
    className="Switch-input"
    disabled={false}
    onChange={[Function]}
    type="checkbox"
  />
  <span
    className="Switch-wrapper Switch-wrapper--regular Switch-wrapper--checked"
  />
</div>
`;

exports[`Switch component 
  checked: true, disabled: true
 1`] = `
<div
  className="Switch Switch--disabled Switch--regular"
>
  <input
    checked={true}
    className="Switch-input"
    disabled={true}
    onChange={[Function]}
    type="checkbox"
  />
  <span
    className="Switch-wrapper Switch-wrapper--disabled Switch-wrapper--regular Switch-wrapper--checked Switch-wrapper--checkedDisabled"
  />
</div>
`;

exports[`Switch component 
  checked: true, size: "regular"
 1`] = `
<div
  className="Switch Switch--regular"
>
  <input
    checked={true}
    className="Switch-input"
    onChange={[Function]}
    type="checkbox"
  />
  <span
    className="Switch-wrapper Switch-wrapper--regular Switch-wrapper--checked"
  />
</div>
`;

exports[`Switch component 
  checked: true, size: "tiny"
 1`] = `
<div
  className="Switch Switch--tiny"
>
  <input
    checked={true}
    className="Switch-input"
    onChange={[Function]}
    type="checkbox"
  />
  <span
    className="Switch-wrapper Switch-wrapper--tiny Switch-wrapper--checked"
  />
</div>
`;
